internal/pkg/propagation/propagation.go

```diff
@@ -257,16 +257,18 @@ func (prop *Propagation) visitCall(call *ssa.Call, maxInstrReached map[*ssa.Basi
 	// Do traverse if the call is being reached through a tainted argument.
 	// Source methods that return tainted values regardless of their arguments should be identified by the fieldpropagator analyzer.
 	if recv := call.Call.Signature().Recv(); recv != nil && sourcetype.IsSourceType(prop.config, prop.taggedFields, recv.Type()) {
-		visitingFromArg := false
-		// Interface types cannot be sources. If the receiver is not a source, the
-		// above condition will be false, so this code won't be executed.
-		// When the receiver's type is statically known (it isn't an interface type),
-		// it will be the first element of the Args slice.
+		// If the receiver is not statically known (it has interface type) and the
+		// method has no arguments, Args will be empty.
 		if len(call.Call.Args) == 0 {
```

Comment:
Consider:

```golang
type SourceProducer interface {  // configure as source
  Produce string
}

func TestProduction(sp SourceProducer) {
  core.Sink(sp.Produce())  // TODO want ".*"
}
```

If we're wanting to permit identification of sources via interface, we need to not skip `len(call.Call.Args) == 0` entirely.

Reply:
I agree, but I think this needs to be handled by the `fieldpropagator` analyzer. At least, that's how it works for `structs`. This analyzer avoids source methods, and then we start new traversals from fieldpropagators. I am planning on implementing this in a future PR. WDYT?

Reply:
(This is what I meant in my comment in the other thread, sorry if that wasn't clear enough.)

> Also, I note that the fieldpropagator's functionality should perhaps be extended to include non-struct types, as currently it will only identify method propagators that return a field. (In a future PR.)

Reply:
Hmmm... this still feels like a hole to me.

If we want `fieldpropogator` to identify certain methods as returning sources, I'm fine with that.  If we want the logic here to not propagate through a value already identified as a source, I'm okay with that.  But I don't like something that makes the assumption that it will be caught elsewhere.  I think we've butted against this before, where we struggle with propagating taint through non-source fields of source-types.

If we're here for optimization, I think it's premature optimization.  If we're skipping here because `sp.Produce()` should be identified as a source, we should be checking for source directly rather than implicitly.

Maybe this is all something that requires a reexamination of `fieldpropagator` though.

Reply:
> But I don't like something that makes the assumption that it will be caught elsewhere.

I agree with your sentiment. However, I think that in the current state, for consistency, we should delegate to the fieldpropagator analyzer. I see too ways that methods can propagate taint: 1. the method is called with a tainted argument (the propagation code handles this), 2. the method always returns a tainted value, regardless of its arguments (the fieldpropagator analyzer handles this). If the method always returns a tainted value, then it is a source, and will be identified as such. 

> If we're here for optimization ...

Optimization was definitely not my intent, and I don't think either option has a significant impact on performance.

> Maybe this is all something that requires a reexamination of fieldpropagator though.

I agree.

Reply:
> I see too [sic] ways that methods can propagate [...]

I see a third: if the value is tainted, every method should propagate taint, not just those identified by `fieldpropagator`.  This is the case that makes me uncomfortable here.

```golang

type FooFace interface{
  Foo() interface{}
}

type Foo struct {
  Valfoo string
}

func (f Foo) Foo() interface{} {
  return f.Valfoo
}

func MakeFoo(s Source) FooFace {
  return Foo{s}
}

func ContaminateInterface(s Source) {
  foo := MakeFoo(s)  // foo should be tainted by s
  val := foo.Foo()  // that taint should propagate to val, but I believe would fail to do so here.
  core.sink(val)
}
```

If I recall, this is an issue with structs right now, too, is it not?  If `s.innocentData` becomes contaminated, we don't necessarily propagate that through `s.GetInnocentData()`?  That's what I was trying to get at when I said

>  I think we've butted against this before, where we struggle with propagating taint through non-source fields of source-types.

------

> I think that in the current state, for consistency, we should delegate to the fieldpropagator analyzer

A proper delegation would work for me, but we're not calling `IsFieldPropagator(call)` here.  (We can't without introducing a circular import, but that's wandering into weeds out of scope.)  We're not delegating, we're assuming that it will have been caught elsewhere.

------

I guess part of the issue here is that we don't have a good definition of what it means for an interface value to be a source.  In the context of a struct, we have the idea that the field `Source.Data` can be a source, so `Source` is a source type holding it.  If `SourceInterface` is a source, is every method a propagator?  Is that what you have in mind for how `fieldpropagator` should handle interfaces?  Or if we are specifying interface-types as a source, does that mean we need to specify which specific methods are the source-y methods?

`</ramble>`

Reply:
In the example you give, if `FooFace` is a source, then I would want the `fieldpropagator` analyzer to flag all of its methods as source-producing (currently, this is not the case; also, `field*` no longer feels like an appropriate name). I think this is the best we can do for now, for interfaces (and likely for other named types, e.g. for a named string I think it is reasonable to assume that the methods will propagate taint).

If `FooFace` is not a source, then currently there won't be an issue because the propagation code only skips methods on sources. I've tested it locally and it works.

> We're not delegating, we're assuming that it will have been caught elsewhere.

Maybe my choice of words was poor. In any case, this is the way that things are working right now. We don't propagate when the receiver is a source, unless one of the arguments is tainted. If the receiver is a source, and the `fieldpropagator` analyzer determined that it propagates taint, then we start a new traversal.

> If SourceInterface is a source, is every method a propagator?

Yes, that feels to me like the most pragmatic stance.

In practice, we haven't seen any interface sources yet, but we're also not preventing users from doing that, and I see how it might make sense for some users. I'm just trying to support this in a way that is consistent with what we have already, so that if we want to handle this differently later, perhaps when we have more information, it is easy to spot the relevant pieces.